app/src/main/java/io/homeassistant/companion/android/sensors/CarSensorManager.kt

```diff
@@ -0,0 +1,306 @@
+            deviceClass = "distance"
+        )
+
+        private val sensorsList = listOf(
+            batteryLevel,
+            carName,
+            carStatus,
+            fuelLevel,
+            odometerValue
+        )
+
+        // track if we have already sent the "open app" message for each sensor
+        private val connectList = mutableMapOf(
+            batteryLevel to false,
+            carName to false,
+            carStatus to false,
+            fuelLevel to false,
+            odometerValue to false
+        )
+
+        private enum class Listener {
+            ENERGY, MODEL, MILEAGE, STATUS,
+        }
+
+        private val listenerSensors = mapOf(
+            Listener.ENERGY to listOf(batteryLevel, fuelLevel),
+            Listener.MODEL to listOf(carName),
+            Listener.STATUS to listOf(carStatus),
+            Listener.MILEAGE to listOf(odometerValue)
+        )
+        private val listenerLastRegistered = mutableMapOf(
+            Listener.ENERGY to -1L,
+            Listener.MODEL to -1L,
+            Listener.STATUS to -1L,
+            Listener.MILEAGE to -1L
+        )
+    }
+
+    override val name: Int
+        get() = R.string.sensor_name_car
+
+    override suspend fun getAvailableSensors(context: Context): List<SensorManager.BasicSensor> {
+        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            sensorsList
+        } else {
+            emptyList()
+        }
+    }
+
+    override fun hasSensor(context: Context): Boolean {
+        // TODO: show sensors for automotive (except odometer) once
+        //  we can ask for special automotive permissions in requiredPermissions
+        return Build.VERSION.SDK_INT >= Build.VERSION_CODES.O &&
+            !context.packageManager.hasSystemFeature(PackageManager.FEATURE_AUTOMOTIVE) &&
+            BuildConfig.FLAVOR == "full"
+    }
+
+    override fun requiredPermissions(sensorId: String): Array<String> {
+        return when {
+            (sensorId == fuelLevel.id || sensorId == batteryLevel.id) -> {
+                arrayOf("com.google.android.gms.permission.CAR_FUEL")
+            }
+            sensorId == odometerValue.id -> {
+                arrayOf("com.google.android.gms.permission.CAR_MILEAGE")
+            }
+            else -> emptyArray()
+        }
+    }
+
+    private lateinit var context: Context
+
+    private fun allDisabled(): Boolean = sensorsList.none { isEnabled(context, it) }
+
+    @RequiresApi(Build.VERSION_CODES.O)
+    private fun connected(): Boolean = HaCarAppService.carInfo != null
+
+    override fun requestSensorUpdate(context: Context) {
+        this.context = context.applicationContext
+        if (allDisabled()) {
+            return
+        }
+
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
+            if (connected()) {
+                updateCarInfo()
+            } else {
+                connectList.forEach { (sensor, alreadySentMessage) ->
```

Comment:
I think the variable `alreadySentMessage` is safe to remove now?

Reply:
I still use it to make sure that we call `onSensorUpdated` only once per sensor when the state is disconnected

Reply:
Other parts of the codebase will take care of not sending the sensor data again if you use a state that is the same as the previous one, so I think it can be removed?

Reply:
Oh I didn't think that was the case, then yes we can remove this